### ISSUE TITLE
op_private assertion on OP_QR via smartmatch

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14513]
Migrated from [rt.perl.org#123848](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123848) (status was 'resolved')

Searchable as RT123848$

**Comment by p5pRT** [comment]
### From @hvds

AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl/>\) finds&#8203;:

% \./miniperl \-e 'my $\_; m// ~~ 0'
Use of my $\_ is experimental at \-e line 1\.
Smartmatch is experimental at \-e line 1\.
miniperl&#8203;: op\.c&#8203;:721&#8203;: Perl\_op\_free&#8203;: Assertion \`\!\(o\->op\_private & ~PL\_op\_private\_valid\[type\]\)' failed\.
Aborted \(core dumped\)
% 

Perl\_ck\_match sets OPpTARGET\_MY while it's still an OP\_MATCH\, but Perl\_ck\_smartmatch then does an inplace CHANGE\_TYPE to OP\_QR\.

Hugo

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Feb 16 07&#8203;:37&#8203;:27 2015\, hv wrote&#8203;:
> AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl/>\) finds&#8203;:
> 
> % \./miniperl \-e 'my $\_; m// ~~ 0'
> Use of my $\_ is experimental at \-e line 1\.
> Smartmatch is experimental at \-e line 1\.
> miniperl&#8203;: op\.c&#8203;:721&#8203;: Perl\_op\_free&#8203;: Assertion \`\!\(o\->op\_private &
> ~PL\_op\_private\_valid\[type\]\)' failed\.
> Aborted \(core dumped\)
>  %
> 
> Perl\_ck\_match sets OPpTARGET\_MY while it's still an OP\_MATCH\, but
> Perl\_ck\_smartmatch then does an inplace CHANGE\_TYPE to OP\_QR\.

Fixed in 57db463f6\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Mar 01\, 2015 at 02&#8203;:14&#8203;:58PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Mon Feb 16 07&#8203;:37&#8203;:27 2015\, hv wrote&#8203;:
> > AFL \(\<http&#8203;://lcamtuf\.coredump\.cx/afl/>\) finds&#8203;:
> > 
> > % \./miniperl \-e 'my $\_; m// ~~ 0'
> > Use of my $\_ is experimental at \-e line 1\.
> > Smartmatch is experimental at \-e line 1\.
> > miniperl&#8203;: op\.c&#8203;:721&#8203;: Perl\_op\_free&#8203;: Assertion \`\!\(o\->op\_private &
> > ~PL\_op\_private\_valid\[type\]\)' failed\.
> > Aborted \(core dumped\)
> >  %
> > 
> > Perl\_ck\_match sets OPpTARGET\_MY while it's still an OP\_MATCH\, but
> > Perl\_ck\_smartmatch then does an inplace CHANGE\_TYPE to OP\_QR\.
> 
> Fixed in 57db463f6\.

&nbsp;    commit 57db463f603fcd67bf53a35aa482bb24aa2c6950
&nbsp;    Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
&nbsp;    Date&#8203;:   Sun Mar 1 14&#8203;:01&#8203;:27 2015 \-0800

&nbsp;    \[perl \#123848\] Allow OPpTARGET\_MY on qr

&nbsp;    // in the scope of lexical $\_ gets the OPpTARGET\_MY flag set\.  If it
&nbsp;    is used as an operand to smartmatch\, it gets converted to a qr op\.
&nbsp;    There is no need to turn off the flag\, since it is harmless at run
&nbsp;    time\.  But we need to allow the flag on this op type to avoid asser\-
&nbsp;    tion failures when it is freed\.

Is that wise? Unless OPpTARGET\_MY is meaningful on OP\_QR\, we should
surely clear it\. At best it can confuse \("why isn't qr flagged as 'T'
in regen/opcodes" etc\)\, and at worst its unnecessarily consumimg a flag
bit\.

\-\- 
Never work with children\, animals\, or actors\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Mar 02 08&#8203;:45&#8203;:09 2015\, davem wrote&#8203;:
>     commit 57db463f603fcd67bf53a35aa482bb24aa2c6950
>     Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>     Date&#8203;:   Sun Mar 1 14&#8203;:01&#8203;:27 2015 \-0800
> 
>     \[perl \#123848\] Allow OPpTARGET\_MY on qr
> 
>     // in the scope of lexical $\_ gets the OPpTARGET\_MY flag set\.  If it
>     is used as an operand to smartmatch\, it gets converted to a qr op\.
>     There is no need to turn off the flag\, since it is harmless at run
>     time\.  But we need to allow the flag on this op type to avoid asser\-
>     tion failures when it is freed\.
> 
> Is that wise? Unless OPpTARGET\_MY is meaningful on OP\_QR\, we should
> surely clear it\.

But clearing flag bits unnecessarily is a waste of CPU cycles\.

> At best it can confuse \("why isn't qr flagged as 'T'
> in regen/opcodes" etc\)\,

There is already a comment that would prevent that particular question\.

> and at worst its unnecessarily consumimg a flag
> bit\.

There are still 6 left\.  In the case of qr\, I think the likelihood of this causing problems is so slim it’s not worth worrying about\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Mar 02\, 2015 at 09&#8203;:52&#8203;:46PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Mon Mar 02 08&#8203;:45&#8203;:09 2015\, davem wrote&#8203;:
> >     commit 57db463f603fcd67bf53a35aa482bb24aa2c6950
> >     Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> >     Date&#8203;:   Sun Mar 1 14&#8203;:01&#8203;:27 2015 \-0800
> > 
> >     \[perl \#123848\] Allow OPpTARGET\_MY on qr
> > 
> >     // in the scope of lexical $\_ gets the OPpTARGET\_MY flag set\.  If it
> >     is used as an operand to smartmatch\, it gets converted to a qr op\.
> >     There is no need to turn off the flag\, since it is harmless at run
> >     time\.  But we need to allow the flag on this op type to avoid asser\-
> >     tion failures when it is freed\.
> > 
> > Is that wise? Unless OPpTARGET\_MY is meaningful on OP\_QR\, we should
> > surely clear it\.
> 
> But clearing flag bits unnecessarily is a waste of CPU cycles\.

But a trivial quantity of cycles and only at compile time\.

> > At best it can confuse \("why isn't qr flagged as 'T'
> > in regen/opcodes" etc\)\,
> 
> There is already a comment that would prevent that particular question\.

There is? I can't see anything obvious in regen/opcodes\, regen/op\_private
or near pp\_qr\(\)\.

> > and at worst its unnecessarily consumimg a flag
> > bit\.
> 
> There are still 6 left\.  In the case of qr\, I think the likelihood of
> this causing problems is so slim it’s not worth worrying about\.

True in this particular case\, but not in general\. And my email was
intended \(perhaps not clearly\) to address the general issue\, not just this
one op\.

\-\- 
You're only as old as you look\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Mar 03 02&#8203;:18&#8203;:20 2015\, davem wrote&#8203;:
> On Mon\, Mar 02\, 2015 at 09&#8203;:52&#8203;:46PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> > But clearing flag bits unnecessarily is a waste of CPU cycles\.
> 
> But a trivial quantity of cycles and only at compile time\.

But you know I’m a fan of microöptimisation\. :\-\)

> > > At best it can confuse \("why isn't qr flagged as 'T'
> > > in regen/opcodes" etc\)\,
> > 
> > There is already a comment that would prevent that particular question\.
> 
> There is? I can't see anything obvious in regen/opcodes\, regen/op\_private
> or near pp\_qr\(\)\.

&nbsp;    \# This flag is also used to indicate matches against implicit $\_\,
&nbsp;    \# where $\_ is lexical; e\.g\. my $\_; \.\.\.\.; /foo/
&nbsp;    qw\(match subst pushre qr trans transr\);

It may not be precise here\, with regard to pushre and qr\, but I think it makes it clear that the flag here does not serve its usual ‘T’ meaning on these ops\.

> > > and at worst its unnecessarily consumimg a flag
> > > bit\.
> > 
> > There are still 6 left\.  In the case of qr\, I think the likelihood of
> > this causing problems is so slim it’s not worth worrying about\.
> 
> True in this particular case\, but not in general\. And my email was
> intended \(perhaps not clearly\) to address the general issue\, not just this
> one op\.

For the general case\, I do think it is fine to leave the flags on\.  We are already doing a fairly good job of avoiding conflicts\, especially thanks to your op\_private work\.  Now\, while there have been bugs resulting from flags left on\, turning them off piecemeal here and there would not have avoided the problem\.

In short\, I’m satisfied with the status quo\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this ticket\.

The issue should now be resolved with the release today of Perl v5\.22\, which is available at http&#8203;://www\.perl\.org/get\.html
\-\- 
Karl Williamson for the Perl 5 team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
